fabtests/component/sock_test.c

```diff
@@ -0,0 +1,347 @@
+#include <sys/time.h>
+#include <netdb.h>
+#include <time.h>
+#include <poll.h>
+#include <sys/epoll.h>
+#include <errno.h>
+
+#include <shared.h>
+
+int *clients;
+
+fd_set select_set;
+int max_sock;
+struct pollfd *poll_set;
+struct epoll_event *events;
+int efpd;
+
+static int start_server(int connections)
+{
+	struct addrinfo *ai = NULL;
+	int ret;
+	int optval = 1;
+
+	ret = getaddrinfo(opts.src_addr, opts.src_port, NULL, &ai);
+	if (ret) {
+		printf("getaddrinfo failed\n");
+		goto err;
+	}
+
+	listen_sock = socket(ai->ai_family, SOCK_STREAM, 0);
+	if (listen_sock < 0) {
+		printf("socket error: %i: %s\n", errno, strerror(errno));
+		ret = -1;
+		goto err;
+	}
+
+	ret = setsockopt(listen_sock, SOL_SOCKET, SO_REUSEADDR, (char *) &optval, 
+		sizeof(optval));
+	if (ret) {
+		printf("error setting socket options\n");
+		goto err;
+	}
+
+	ret = bind(listen_sock, ai->ai_addr, ai->ai_addrlen);
+	if (ret) {
+		printf("bind failed\n");
+		goto err;
+	}
+
+	ret = listen(listen_sock, connections);
+	if (ret) {
+		printf ("listen failed\n");
+		goto err;
+	}
+
+	freeaddrinfo(ai);
+	return ret;
+
+err:
+	close(listen_sock);
+	freeaddrinfo(ai);
+	return ret;
+}
+
+static int server_connect(int connections)
+{
+	int new_sock, i;
+
+	for (i = 0; i < connections; i++) {
+		new_sock = accept(listen_sock, NULL, NULL);
+		if (new_sock < 0) {
+			printf("error during server init\n");
+			return -1;
+		}
+
+		clients[i] = new_sock;
+		
+	}
+	return 0;
+}
+
+static int client_connect(int connections)
+{
+	struct addrinfo *res;
+	int i, ret, new_sock;
+
+	for (i = 0; i < connections; i++) {
+
+		ret = getaddrinfo(opts.dst_addr, opts.dst_port, NULL, &res);
+		if (ret) {
+			printf("getaddrinfo failed\n");
+			goto err;
+		}
+
+		new_sock = socket(res->ai_family, SOCK_STREAM, 0);
+		if (new_sock < 0) {
+			ret = -1;
+			goto err;
+		}
+
+		ret = connect(new_sock, res->ai_addr, res->ai_addrlen);
+		if (ret)
+			goto err;
+
+		clients[i] = new_sock;
+	}
+
+err:
+	freeaddrinfo(res);
+	return ret;
+}
+
+static int init_select() 
+{
+	FD_ZERO(&select_set);
+	FD_SET(listen_sock, &select_set);
+	max_sock = listen_sock;
+
+	return 0;
+}
+
+static int time_select(int connections) 
+{
+	int i;
+	struct timeval timeout;
+	timeout.tv_sec = 0;
+	timeout.tv_usec = 0;
+
+	ft_start();
+	for (i = 0; i < connections; i++) {
+		FD_SET(clients[i], &select_set);
+
+		if (clients[i] > max_sock)
+			max_sock = clients[i];
+	}
+
+	select(max_sock + 1, &select_set, NULL, NULL, &timeout);
+	ft_stop();
+
+	printf("%i: %ld\n", connections, (end.tv_nsec + end.tv_sec * 1000000000) - 
+		(start.tv_nsec + start.tv_sec * 1000000000));
+	return 0;
+}
+
+static int init_poll(int connections)
+{
+	int i;
+	poll_set = malloc(sizeof(struct pollfd) * connections);
+	if (!poll_set)
+		return -1;
+
+	for (i = 0; i < connections; i++) {
+		poll_set[i].fd = clients[i];
+	}
+
+	return 0;
+}
+
+static int time_poll(int connections)
+{
+	ft_start();
+	poll(poll_set, POLLIN, 0);
+	ft_stop();
```

Comment:
Picking on one of the time_xxx calls.  Trying to time a single instance of a call (select, poll, epoll) isn't going to be accurate enough to be useful.  We need to time several thousand calls and average that.  E.g.
```
ft_start()
for (i = 0; i < 100000; i++)
    time_poll();
ft_stop()
print_results()

ft_start()
for (i = 0; i < 100000; i+)
    time_select()
ft_stop()
```